File: leetcode/10-Search-Insert-Position/python/main.py
class Solution:
    def searchInsert(self, nums, target: int) -> int:
        head_index = 0
        tail_index = len(nums) - 1
        result = 0

        while head_index <= tail_index:
            mid_index = (head_index + tail_index) // 2

            if nums[mid_index] == target:
                result = mid_index
                break

            if nums[head_index] <= target < nums[mid_index]:
                # 左側にあるので次のループは左
                tail_index = mid_index - 1
            elif nums[mid_index] < target <= nums[tail_index]:
                # 右側にあるので次のループは右
                head_index = mid_index + 1
            elif target < nums[head_index]:
                result = head_index
                break
            elif nums[tail_index] < target:
                result = tail_index + 1
                break

        return result

File: leetcode/10-Search-Insert-Position/go/main.go
package main

import "fmt"

func searchInsert(nums []int, target int) int {
	left, right := 0, len(nums)-1
	var mid int
	result := -1

	fmt.Println()
	for left <= right {
		mid = (left + right) / 2 // int型なので切り捨て
		fmt.Printf("Start left: %v, mid: %v, right: %v\n", left, mid, right)

		if nums[mid] == target {
			result = mid
			break
		}

		if target < nums[left] {
			// 今のグループの左側にある
			result = left
			break
		} else if nums[right] < target {
			// 今のグループの右側にある
			result = right + 1
			break
		}

		if nums[left] <= target && target < nums[mid] {
			// 左側のグループにある
			right = mid - 1
		} else {
			// 右側のグループにある
			left = mid + 1
		}
		fmt.Printf("End left: %v, mid: %v, right: %v\n", left, mid, right)

	}

	return result
}
